frontend/src/stimulus/controllers/dynamic/work-packages/activities-tab/services/dom_helpers.ts

```diff
@@ -0,0 +1,98 @@
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program; if not, write to the Free Software
+* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+*
+* See COPYRIGHT and LICENSE files for more details.
+* ++
+*/
+
+export namespace DomHelpers {
+  /**
+   * Preserves scroll position during DOM manipulations that change content height.
+   *
+   * This is crucial for infinite scroll UX - when older content is prepended above
+   * the current view, we need to adjust scroll position so the user doesn't suddenly
+   * jump to a different part of the content.
+   *
+   * @param {HTMLElement} container - The scrollable container
+   * @param {boolean} top - Whether content is being added at the top (prepend)
+   *                         true: adjust scroll for prepended content
+   *                         false: maintain position for appended content
+   * @param {Function} fn - Async function that performs the DOM manipulation
+   *
+   * Algorithm:
+   * - Capture current scroll position and total height BEFORE DOM changes
+   * - Execute the DOM manipulation (adding/removing content)
+   * - Calculate height difference and adjust scroll position accordingly
+   *
+   * For prepend: scrollTop + heightDifference (push view down by added content height)
+   * For append: maintain original scrollTop (new content below doesn't affect view)
+   */
+  export async function keepScroll(container:HTMLElement, top:boolean, fn:() => Promise<void>) {
+    pauseInertiaScroll(container);
+
+    const scrollTop = container.scrollTop;
+    const scrollHeight = container.scrollHeight;
+
+    await fn();
+
+    if (top) {
+      container.scrollTop = scrollTop + (container.scrollHeight - scrollHeight);
+    } else {
+      container.scrollTop = scrollTop;
+    }
+  }
+
+  /**
+   * Temporarily pauses inertial/momentum scrolling to prevent scroll position drift.
+   *
+   * CRITICAL for iOS Safari and other mobile browsers with momentum scrolling!
+   *
+   * The Problem:
+   * When programmatically setting scrollTop during an active momentum scroll,
+   * the browser continues the inertial scroll AFTER our position adjustment,
+   * causing the scroll position to drift away from our intended target.
+   *
+   * The Solution:
+   * - Set overflow: hidden to immediately stop any momentum scrolling
+   * - Use requestAnimationFrame to restore overflow on next frame
+   * - This gives us a clean slate to set the exact scroll position
+   *
+   * Without this, scroll preservation would be unreliable on mobile devices,
+   * especially during infinite scroll when users are actively scrolling up.
+   *
+   * Credit: This technique was adapted from Basecamp's Campfire implementation
+   * where they solved similar scroll position issues in their real-time chat.
```

Comment:
IANAL so check with @wielinde and team - However I think it's good practice to also include the original Basecamp LICENSE in this header.

Reply:
Ah, good point- it's MIT Licensed and OSS. I'll add these for reference.

https://github.com/basecamp/once-campfire/blob/5c0526eaf7f83d129472a520fa673bfcb9c2f397/MIT-LICENSE

https://github.com/basecamp/once-campfire/blob/5c0526eaf7f83d129472a520fa673bfcb9c2f397/app/javascript/helpers/dom_helpers.js#L44-L65